src/Analyzers/CSharp/Analyzers/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp.UsePatternCombinators
+{
+    using static AnalyzedPattern;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpUsePatternCombinatorsDiagnosticAnalyzer :
+        AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpUsePatternCombinatorsDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UsePatternCombinatorsDiagnosticId,
+                CSharpCodeStyleOptions.PreferPatternMatching,
+                LanguageNames.CSharp,
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_pattern_matching), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeNode,
+                SyntaxKind.LogicalAndExpression,
+                SyntaxKind.LogicalOrExpression,
+                SyntaxKind.LogicalNotExpression);
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var expression = (ExpressionSyntax)context.Node;
+
+            if (!IsTopmostExpression(expression))
+                return;
+
+            var syntaxTree = expression.SyntaxTree;
+            var cancellationToken = context.CancellationToken;
+
+            if (!((CSharpParseOptions)syntaxTree.Options).LanguageVersion.IsCSharp9OrAbove())
+                return;
+
+            var styleOption = context.Options.GetOption(CSharpCodeStyleOptions.PreferPatternMatching, syntaxTree, cancellationToken);
+            if (!styleOption.Value)
+                return;
+
+            var operation = context.SemanticModel.GetOperation(expression, cancellationToken);
+            if (operation is null)
+                return;
+
+            var pattern = CSharpUsePatternCombinatorsAnalyzer.Analyze(operation);
+            if (pattern is null)
+                return;
+
+            // Avoid rewriting trivial patterns, such as a single relational or a constant pattern.
+            if (IsTrivial(pattern))
+                return;
+
+            // C# 9.0 does not support pattern variables under `not` and `or` combinators.
```

Comment:
note: i believe for 9.0 we will allow:  `x is not string y`.  specifically we will allow a `not` directly under an `is` that has a pattern which declares a variable directly under it.  @gafter  to confirm.

Reply:
Yes, that is the plan for C# 9.0. It is implemented in https://github.com/dotnet/roslyn/pull/44513 but the compiler team is fully heads down on records work and might not review it in time for it to go into the upcoming preview.

Reply:
Thanks for letting us know @gafter !

Reply:
>might not review it in time for it to go into the upcoming preview.

To avoid confusion, I think we can leave this as-is for now.